pkg/dns/aws/dns_test.go

```diff
@@ -0,0 +1,264 @@
+	}))
+	sess.Handlers.Send.Clear()
+	sess.Handlers.Unmarshal.Clear()
+	sess.Handlers.UnmarshalMeta.Clear()
+	sess.Handlers.ValidateResponse.Clear()
+	sess.Handlers.Unmarshal.PushBack(func(r *request.Request) {
+		r.Data = resps[reqNum]
+		reqNum++
+	})
+
+	config := &aws.Config{
+		MaxRetries: aws.Int(1),
+		Region:     aws.String("ap-southeast-2"),
+	}
+
+	route53Config := sess.ClientConfig("route53", config)
+	route53 := &route53.Route53{
+		Client: client.New(
+			*route53Config.Config,
+			metadata.ClientInfo{
+				ServiceName:   "route53",
+				ServiceID:     "Route 53",
+				SigningName:   "",
+				SigningRegion: route53Config.SigningRegion,
+				Endpoint:      route53Config.Endpoint + "/route53",
+				APIVersion:    "2013-04-01",
+				JSONVersion:   "1.1",
+				TargetPrefix:  "MockServer",
+			},
+			route53Config.Handlers,
+		),
+	}
+
+	elbConfig := sess.ClientConfig("elasticloadbalancing", config)
+	elb := &elb.ELB{
+		Client: client.New(
+			*elbConfig.Config,
+			metadata.ClientInfo{
+				ServiceName:   "elasticloadbalancing",
+				ServiceID:     "Elastic Load Balancing",
+				SigningName:   "",
+				SigningRegion: elbConfig.SigningRegion,
+				Endpoint:      elbConfig.Endpoint + "/elb",
+				APIVersion:    "2012-06-01",
+				JSONVersion:   "1.1",
+				TargetPrefix:  "MockServer",
+			},
+			elbConfig.Handlers,
+		),
+	}
+
+	taggingConfig := sess.ClientConfig("resourcegroupstaggingapi", config)
+	tagging := &resourcegroupstaggingapi.ResourceGroupsTaggingAPI{
+		Client: client.New(
+			*taggingConfig.Config,
+			metadata.ClientInfo{
+				ServiceName:   "tagging",
+				ServiceID:     "Resource Groups Tagging API",
+				SigningName:   "",
+				SigningRegion: taggingConfig.SigningRegion,
+				Endpoint:      taggingConfig.Endpoint + "/tagging",
+				APIVersion:    "2017-01-26",
+				JSONVersion:   "1.1",
+				TargetPrefix:  "ResourceGroupsTaggingAPI_20170126",
+			},
+			taggingConfig.Handlers,
+		),
+	}
+
+	return &Provider{
+		elb:            elb,
+		route53:        route53,
+		tags:           tagging,
+		idsToTags:      map[string]map[string]string{},
+		lbZones:        map[string]string{},
+		updatedRecords: sets.NewString(),
+	}
+}
+
+func TestEnsureAWSDNS(t *testing.T) {
+	var resps []interface{}
+	enqueueResponse := func(resp interface{}) {
+		resps = append(resps, resp)
+	}
+
+	record := iov1.DNSRecord{
+		Spec: iov1.DNSRecordSpec{
+			DNSName:    "subdomain.dnszone.io.",
+			RecordType: iov1.CNAMERecordType,
+			Targets:    []string{"afcfa1e69205711e99e9906f0636dbcb-2044172706.ap-southeast-2.elb.amazonaws.com"},
+			RecordTTL:  120,
+		},
+	}
+
+	dnsZone := configv1.DNSZone{
+		ID: "Z1GM3OXH4ZPM65",
+	}
+
+	enqueueResponse(&elb.DescribeLoadBalancersOutput{
+		LoadBalancerDescriptions: []*elb.LoadBalancerDescription{
+			&elb.LoadBalancerDescription{
+				DNSName:                   &record.Spec.Targets[0],
+				CanonicalHostedZoneNameID: &dnsZone.ID,
+			},
+		},
+	})
+	enqueueResponse(&route53.ChangeResourceRecordSetsInput{})
```

Comment:
I'm not sure the `enqueueResponse` helper is really improving readability.  Probably better to declare the responses and then declare `resps`:
```go
	elbResponse := &elb.DescribeLoadBalancersOutput{
￼		LoadBalancerDescriptions: []*elb.LoadBalancerDescription{
￼			&elb.LoadBalancerDescription{
￼				DNSName:                   &record.Spec.Targets[0],
￼				CanonicalHostedZoneNameID: &dnsZone.ID,
￼			},
￼		},
￼	}
	route53Response := &route53.ChangeResourceRecordSetsInput{}
	responses := []interface{elbResponse, route53Response}
	mgr := FakeAWSProvider(responses)
```
What do you think?

Also, what do you think of my earlier suggestion to refactor these tests into a table-driven test (https://github.com/openshift/cluster-ingress-operator/commit/a3b29e2e4598581dea2a1d2a9789828f2bf76a8f#r35977141)?